resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/KubernetesSubmissionStepsOrchestrator.scala

```diff
@@ -0,0 +1,135 @@
+  // names of the Kubernetes resources from e.g. Kubectl or the Kubernetes dashboard to the
+  // application the user submitted. However, we can't use the application name in the label, as
+  // label values are considerably restrictive, e.g. must be no longer than 63 characters in
+  // length. So we generate a separate identifier for the app ID itself, and bookkeeping that
+  // requires finding "all pods for this application" should use the kubernetesAppId.
+  private val kubernetesResourceNamePrefix =
+      s"$appName-$launchTime".toLowerCase.replaceAll("\\.", "-")
+  private val jarsDownloadPath = submissionSparkConf.get(INIT_CONTAINER_JARS_DOWNLOAD_LOCATION)
+  private val filesDownloadPath = submissionSparkConf.get(INIT_CONTAINER_FILES_DOWNLOAD_LOCATION)
+  private val dockerImagePullPolicy = submissionSparkConf.get(DOCKER_IMAGE_PULL_POLICY)
+  private val initContainerConfigMapName = s"$kubernetesResourceNamePrefix-init-config"
+
+  def getAllSubmissionSteps(): Seq[KubernetesSubmissionStep] = {
+    val additionalMainAppJar = mainAppResource match {
+      case JavaMainAppResource(resource) if resource != SparkLauncher.NO_RESOURCE =>
+        Option(resource)
+      case _ => Option.empty
+    }
+    val additionalMainAppPythonFile = mainAppResource match {
+      case PythonMainAppResource(resource) if resource != SparkLauncher.NO_RESOURCE =>
+        Option(resource)
+      case _ => Option.empty
+    }
+    val sparkJars = submissionSparkConf.getOption("spark.jars")
+        .map(_.split(","))
+        .getOrElse(Array.empty[String]) ++ additionalMainAppJar.toSeq
+    val sparkFiles = submissionSparkConf.getOption("spark.files")
+        .map(_.split(","))
+        .getOrElse(Array.empty[String]) ++ additionalMainAppPythonFile.toSeq
+    val driverCustomLabels = ConfigurationUtils.combinePrefixedKeyValuePairsWithDeprecatedConf(
+        submissionSparkConf,
+        KUBERNETES_DRIVER_LABEL_PREFIX,
+        KUBERNETES_DRIVER_LABELS,
+        "label")
+    require(!driverCustomLabels.contains(SPARK_APP_ID_LABEL), s"Label with key " +
+        s" $SPARK_APP_ID_LABEL is not allowed as it is reserved for Spark bookkeeping" +
+        s" operations.")
+    val allDriverLabels = driverCustomLabels ++ Map(
+        SPARK_APP_ID_LABEL -> kubernetesAppId,
+        SPARK_ROLE_LABEL -> SPARK_POD_DRIVER_ROLE)
+    val initialSubmissionStep = new BaseSubmissionStep(
+        kubernetesAppId,
+        kubernetesResourceNamePrefix,
+        allDriverLabels,
+        dockerImagePullPolicy,
+        appName,
+        mainClass,
+        appArgs,
+        submissionSparkConf)
+    val kubernetesCredentialsStep = new DriverKubernetesCredentialsStep(
+        submissionSparkConf, kubernetesAppId)
+    val pythonStep = mainAppResource match {
+      case PythonMainAppResource(mainPyResource) =>
+        Option(new PythonStep(mainPyResource, additionalPythonFiles, filesDownloadPath))
+      case _ => Option.empty[KubernetesSubmissionStep]
```

Comment:
So here instead of an Empty Step you would resolve the arguments to just be appArgs as you would have before in the BaseSubmissionStep, but otherwise you use the PythonStep. I believe SparkR has similair arguments to Java/Scala so this generalization could be just called: NonPythonArgumentResolver